### PR TITLE
Fix multi-account password saving

### PR DESCRIPTION
**Pull request by vgaming**
Fixes GH-6288, and hopefully makes the code
more readable / easier to understand.

### CODE REVIEW

**Comment by vgaming** [comment]
I tested the new code locally, it works both for single-account use case and for storing/loading multiple accounts

**Comment by vgaming** [comment]
I also hate to admit, but I cannot pinpoint *where exactly* the old code was broken. In my perception that was just messy code that is really hard to understand, with unneeded additional variable `offset`, triple resizes that adjust for escaped and not escaped keys... A lot of unnecessary complexity. In the issue GH-6288 itself, however, it was observed that `credentials` file is not split correctly, with the `CREDENTIAL_SEPARATOR` obviously broken. I can copy more of the relevant IRC discussions here if there's a question

**Comment by Wedge009** [comment]
As discussed in #wesnoth-dev, it is indeed strange that it didn't expose the issue for multi-account credential saving for you in 1.14. But it does look like replacing manual `std::copy` and `offset` tracking is a good thing and if it solves the bug for you and @Hejnewar.

It looks like only @CelticMinstrel has done any meaningful work on this code, so I'll suggest he review this before merging.

**Comment by vgaming** [comment]
UPD: When using this branch on an old corrupted `credentials` file from wesnoth-1.16.1, the passwords have still to be entered once, but they're saved successfully and for the user it works as expected. (This PR doesn't really change reading of the credentials file.) What remains however is that the `credentials` file will keep those extra broken/unneeded entries, even with the new PR. They'll just stay as garbage around, without use.

**Comment by vgaming** [comment]
To provide a bit of context for people who weren't today in IRC. Without this PR, for single-account credentials file, the entry in credentials file looks like:  server: server.wesnoth.org:15000, login: vasili. For multi-account credentials file, without the PR, corrupted entries will be created upon save, such as:  server: server.wesnoth.org:15000, login: login: vasili@server.wesnoth.org:15000=�����;sa��vasya2   (the login spans much further than it should, including the server and probably password of the previous/next entries)

File: src/preferences/credentials.cpp [hunk 1]
```diff
@@ -230,21 +233,15 @@ namespace preferences
 			filesystem::delete_file(filesystem::get_credentials_file());
 			return;
 		}
-		secure_buffer credentials_data(1, CREDENTIAL_SEPARATOR);
-		std::size_t offset = 1;
+		secure_buffer credentials_data(0, CREDENTIAL_SEPARATOR);
```
Comment (CelticMinstrel) [inline src/preferences/credentials.cpp:236]: `(0, CREDENTIAL_SEPARATOR)` does nothing - just remove it.

**Review by CelticMinstrel — CHANGES_REQUESTED** [review]
Looks good, just one minor change to make.

File: src/preferences/credentials.cpp (hunk 1 above)
Comment (vgaming) [reply src/preferences/credentials.cpp:236]: Hey, I was expecting that -- I just newb enough in C++ to not know the alternative.. So what do I write then, `credentials_data();` or `credentials_data;` ?

Comment (vgaming) [reply src/preferences/credentials.cpp]: UPD: neverming, found it, the compiler is pretty unambiguous on what I should do anyway :sweat_smile:

Comment (vgaming) [reply src/preferences/credentials.cpp]: Thanks for the review! Addressed the comment

**Review by CelticMinstrel — APPROVED** [review]

**Comment by AI0867** [comment]
The only potential bug I see in the old code is that the second resize (for the difference in size between escaped and non-escaped keys) doesn't pad the buffer with CREDENTIAL_SEPARATOR.

The new code looks like a definite improvement. It might be a bit slower due to not pre-allocating the buffer, but this is hardly a performance-critical path.

**Comment by vgaming** [comment]
@AI0867 totally agree. Also, for this code path, I'd prefer readability, maintainability and less bugs, over performance optimization %insert-donald-knuth-quote-here%

**Comment by AI0867** [comment]
Mostly unrelated, but I'm unsure if I should open a new issue for this: I noticed a TODO to do key stretching on the autogenerated key. This seems like a good idea, but unfortunately, it comes with a whole host of backwards-compatibility issues, especially if we want to be able to allow older versions of wesnoth to keep using the same credentials file.

Should we open a new issue for that to think about what a good design would even look like?

**Comment by vgaming** [comment]
@AI0867 I don't know, I can't really say. Normally I always try to prioritize fixing bugs, and only then doing new features. But that is subjective of course.

**Pull request merged**